compiler/hash-source/src/constant.rs

```diff
@@ -0,0 +1,291 @@
+//! Hash source interned constant store. This is a data store
+//! for interning observed constants within the source file of
+//! a program. This is done so that constants don't need to
+//! be passed around each stage of the compiler and can just
+//! be interned, and accessed when needed.
+
+use std::fmt::Display;
+
+use dashmap::DashMap;
+use fnv::FnvBuildHasher;
+use hash_utils::counter;
+use lazy_static::lazy_static;
+use num_bigint::BigInt;
+
+use crate::identifier::{Identifier, CORE_IDENTIFIERS};
+
+/// Interned float constant which stores the value of the float, and
+/// an optional `type ascription` which is a suffix on the literal
+/// describing which float kind it is, either being `f32` or `f64`.
+#[derive(Debug, Clone, Copy)]
+pub struct FloatConstant {
+    /// Raw value of the float
+    pub value: f64,
+    /// If the constant contains a type ascription, as specified
+    /// when the constant is declared, e.g. `32.4f64`
+    pub suffix: Option<Identifier>,
+}
+
+counter! {
+    name: InternedFloat,
+    counter_name: INTERNED_FLOAT_COUNTER,
+    visibility: pub,
+    method_visibility:,
+}
+
+impl Display for FloatConstant {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.value)?;
+
+        if let Some(suffix) = self.suffix {
+            write!(f, "{suffix}")?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Display for InternedFloat {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", CONSTANT_MAP.lookup_float_constant(*self))
+    }
+}
+
+/// Value of the [IntConstant], this kind be either the `inlined`
+/// variant where we just fallback to using `u128` for small sized
+/// integer constants, and then in the unlikely scenario of needing
+/// more than a [u128] to represent the constant, we will then
+/// fallback to [BigInt].
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub enum IntConstantValue {
+    /// For small values, we store inline
+    Small(u128),
```

Comment:
Maybe u64? 99.99% of the time no one will write a literal > 64 bits

Reply:
Also, maybe use an inline byte array instead  `[u8; 8]` (or `[u8; 16]`)? better communicates the fact that this data is "serialised" in the case of negative integers eg.